src/reducers/vis-state-updaters.js

```diff
@@ -1328,3 +1344,167 @@ export function updateAllLayerDomainData(state, dataId, newFilter) {
     layerData: newLayerDatas
   };
 }
+
+/**
+ * Update editor features
+ * @memberof visStateUpdaters
+ * @param {Object} state `visState`
+ * @param {[Object]} features to store
+ * @return {Object} nextState
+ */
+export function setFeaturesUpdater(state, {features = []}) {
+  let newState = {
+    ...state,
+    editor: {
+      ...state.editor,
+      features
+    }
+  };
+
+  // Retrieve existing feature
+  const {selectedFeature} = state.editor;
+
+  if (!selectedFeature) {
+    return newState;
+  }
+
+  // TODO: check if the feature has changed
+  const feature = features.find(f => f.id === selectedFeature.id);
+
+  if (!feature) {
+    return newState;
+  }
+
+  const newFilters = state.filters.map(f => {
+    if (f.type === FILTER_TYPES.polygon && f.feature.id === selectedFeature.id) {
+      return updatePolygonFilter(f, feature)
+    }
+    return f;
+  });
+
+  const filtersIdx = getFiltersIdxByFeatureId(newState, selectedFeature.id);
+
+  // no filters found
+  if (filtersIdx.length === 0) {
+    return newState;
+  }
+
+  const datasets = filtersIdx.reduce((acc, idx) => {
+    const dataId = state.filters[idx].dataId;
+    return {
+      ...acc,
+      [dataId]: {
+        ...state.datasets[dataId],
+        ...filterData(state.datasets[dataId].allData, dataId, newFilters, state.layers)
+      }
+    }
+  }, state.datasets);
+
+  newState = {
+    ...state,
+    // setting filters
+    filters: newFilters,
+    // updating datasets
+    datasets
+  };
+
+  return updateAllLayerDomainData(newState, Object.keys(datasets));
+
+}
+
+/**
+ * Set the current selected feature
+ * @memberof uiStateUpdaters
+ * @param {Object} state `uiState`
+ * @param {[Object]} features to store
+ * @return {Object} nextState
+ */
+export const setSelectedFeatureUpdater = (state, {selectedFeatureId}) => ({
+  ...state,
+  editor: {
+    ...state.editor,
+    selectedFeature: {id: selectedFeatureId}
+  }
+});
+
+/**
+ * Delete existing feature from filters
+ * @memberof visStateUpdaters
+ * @param {Object} state `visState`
+ * @param {string} selectedFeatureId feature to delete
+ * @return {Object} nextState
+ */
+export function deleteFeatureUpdater(state, {featureId}) {
+  if (!featureId) {
+    return state;
+  }
+
+  const {editor} = state;
+
+  // modify editor object
+  const newEditor = {
+    ...editor,
+    features: editor.features.filter(f => f.id !== featureId),
+    selectedFeature: {
+      id: null
+    }
+  };
+
+  const filtersIdx = getFiltersIdxByFeatureId(state, featureId);
+
+  const newState = {
+    ...state,
+    editor: newEditor
+  };
+
+  return filtersIdx.length === 0 ? newState : removeFeatureFilters(newState, filtersIdx);
+}
+
+/**
+ * Toggle feature as layer filter
+ * @memberof visStateUpdaters
+ * @param state
+ * @param {Object} payload
+ * @param {string} payload.featureId
+ * @param {Object} payload.layer
+ * @return {Object} nextState
+ */
+export function togglePolygonFilterUpdater(state, payload) {
```

Comment:
do we need to store `feature` that is also a polygon filter in 2 places? It's making the logic really complicated. and you have to add this `filterData` everywhere in all feature update / delete calls.

I think as soon as someone toggle a polygon into a filter, you can move the `feature` into `polygonFilter.value`. So all the update / remove call on polygon filter can just be using `setFilter` `removeFilter`

Obviously, on the `Editor` side, you will have to assign different `onUpdate`, `onDeleteFeature` based on whether the `selectedFeatureId` is a filter or not. But I think this simplifies the logic in `visStateUpdater` so we don't add this many specific updaters just for polygon feature

Reply:
in filter.value we store the turf.js representation of the feature and not original one so in case a user wants to apply the same feature as filter to multiple layers we still need the original one.
Another reason. If a user has a feature and a filter mapped against and decide to delete/toggle the filter, the feature should still be present on the map

Reply:
Again on this point. 1. turf representation and the original feature is the same. 2. if a user deletes a polygon filter on map, the feature should not be present on the map. I don't think we should store them in 2 places. we would have to implement `getFilterIdxByFeatureId`, `currentFeature = editorFeatures.find(feat => feat.id === featureId);` just to get each one.

Reply:
I tested out in [this commit](https://github.com/keplergl/kepler.gl/commit/7f927840ab976305eee3200d3aedb903ab60c0df#diff-3c42abf857bb61c5f6578c72771405d2R1658)